\section*{Introduction}

Visualising multiple sequence alignments (MSAs) is a common task in
computational biology. Many alignment viewers have a graphical user interface
(GUI) and are hence unsuitable for use on headless or remote systems such as
high-performance computing (HPC) clusters.  Command-line tools do exist, for
example \texttt{alan}\cite{alan}, which stands out as a particularly elegant solution,
since it is built on standard Unix tools such as \texttt{awk} and \texttt{less}
--- indeed, it served as the initial inspiration for the present work. This
means, however, that \texttt{alan}'s interactivity is limited to that of a pager:
features such as zooming, reordering sequences, as well as computing and
displaying a consensus sequence are absent. While
\texttt{showalign}\cite{emboss} can compute a consensus, it does not support
colouring residues, and the user must explicitly call a pager in order to scroll
through the alignment. Other programs like \texttt{alen}\cite{alen} are
interactive, but not all have built-in residue colour schemes or the ability to
visually represent metrics such as similarity to the consensus, or to reorder
sequences according to such metrics. The capacity to fit a large alignment on
screen, typically by only displaying a subset of the sequences and columns, is
also rare. In summary, text-based MSA viewers collectively provide a substantial
range of functions, but no viewer implements all, or even most, of them. In this
work we introduce \texttt{termal}, which combines most of these features in a
single application.

\section*{Interface}

Apart from the alignment sequences, which occupy the main pane, \texttt{termal}
also displays sequence labels and  ordinal numbers, a consensus sequence, and a
conservation bar plot; it also displays sequence metrics such as similarity to
the consensus, or (ungapped) length (figure~\ref{fig:screen}). The alignment can
be scrolled one sequence/column at a time using arrow keys or familiar Vim-style
motion keys \texttt{h}, \texttt{j}, \texttt{k}, and \texttt{l}; similar
keystrokes allow jumping by screenfuls or to the edges of the alignment.

By default, residues of nucleotide alignments are coloured according to
Jalview's\cite{waterhouse2009jalview} nucleotide colour scheme, while protein
alignments use that of ClustalX\cite{larkin2007clustal}. An alternative colour
scheme for protein is Lesk's\cite{lesk2019introduction}, and all alignments can
be rendered in monochrome. 

Alignments that exceed the screen size—horizontally, vertically, or both—can be
zoomed out to display only the first and last row and column, along with a
sample of equidistant intermediates in both dimensions; the sample size is
adjusted to fill the available screen space. This allows regions of high
conservation to be spotted without scrolling. An alternative zoomed-out mode
preserves the alignment's aspect ratio, at the cost of some unused space.

The sequences can be reordered according to the currently-displayed sequence
metric, in increasing or decreasing order. This allows e.g. to group the most
complete sequences together, or those that best match the consensus.

The width of the label pane can be adjusted to fit label length, and both the
side and bottom panes can be hidden to maximise the space allocated to the
alignment.

\texttt{termal} includes a built-in help screen that lists all key bindings.

\begin{figure*}[!t]%
\centering
	\includegraphics[width=\textwidth]{figure-1.pdf}
\caption{%
	A snapshot of the interface of \texttt{termal} v1.0.0, showing a protein alignment. A:
	alignment filename and dimensions, B: sequence numbers pane, C: sequence
	labels pane, D: metric bar plot pane (currently displaying sequence similarity
	with the consensus), E: alignment pane, F: bottom pane, displaying sequence
	position, consensus, and conservation bar plot. \\
	In this example the sequences are in the original (file) order, and use the
	ClustalX colour scheme. The view is zoomed in, that is, only a fraction of the
	alignment is displayed. To reproduce: run \texttt{termal} on the example
	datafile supplied in the distribution.}
	\label{fig:screen}
\end{figure*}


\section*{Comparison with Other \textsc{TUI} Tools}

\paragraph*{Dependencies and Installation}

\texttt{termal} and \texttt{alen} are written in Rust\cite{klabnik2019rust} and
compile to a single binary with no dependencies. By contrast,
\texttt{alv}\cite{Arvestad2018} requires BioPython and \texttt{showalign} is
part of the \textsc{EMBOSS} suite.

\paragraph*{Features}

\texttt{termal} features zooming, interactive navigation, selectable themes and
color maps, as well as ordering according to metrics. This combination of
features is not found in the other programs considered, although all of them
offer at least some of these features.

\section*{Performance and Limitations}

\texttt{termal} has been tested on alignments exceeding 15,000 sequences and
1,500 columns (\textasciitilde{}22 million alignment cells), with startup and
initial rendering completing in under one second on a machine with a
12th-generation Intel\textregistered{} Core\textsuperscript{TM} i5-1240P
processor and 16 GB of RAM running Linux 6.14.2. In practice, interactive
performance is limited more by the speed of the terminal emulator than by
\texttt{termal} itself.  GPU-accelerated terminals such as
Alacritty\cite{alacritty}, Kitty\cite{kitty} and Ghostty\cite{ghostty} offer
smoother scrolling at large screen sizes than do more traditional emulators, and
terminals that support 24-bit colours give better results.  \texttt{termal} is
currently restricted to alignments in Fasta and Stockholm formats.


\section*{Availability}

\texttt{termal} is distributed under the MIT license. It is available as a
precompiled, statically-linked binary (for Linux, MacOS, and Windows), with no external
dependencies or runtime environment required, from
\url{https://github.com/sib-swiss/termal.git}.\linebreak
Alternatively, users with Rust
installed can install it via \texttt{cargo install termal-msa}. The source code
has also been deposited on Zenodo (\url{https://zenodo.org/records/15472432}).

\section*{Conclusion}

While this work is not intended as a comprehensive review of terminal-based
alignment viewers, we surveyed several of them, including \texttt{showalign},
\texttt{alan}, \texttt{alv}, and \texttt{alen}. To our knowledge,
\texttt{termal} combines a unique set of features in a terminal interface.  Its
minimal dependencies and fast startup make \texttt{termal} suitable for both
\textit{ad-hoc} use and for integration into semi-automated workflows requiring
alignments to be reviewed in a terminal. Accordingly, \texttt{termal} fills a
niche for fast, interactive MSA exploration directly in the terminal, making it
an ideal tool for remote bioinformatics workflows.

\section*{Acknowledgements}

The development of \texttt{termal} was funded by Swiss National Science
Foundation BRIDGE Discovery grant 40B2-0\_194701. The author wishes to thank
Drs Guillaume Cailleau and Sébastien Moretti for insightful comments on the
program.

\section*{Author Contributions (CRediT)}
Thomas Junier: Conceptualization, Software, Validation, Visualization Writing – Original
Draft, Writing – Review \& Editing.

\section*{Conflicts of Interest}

The author declares no conflict of interest.

%\bibliographystyle{natbib} % doesn't seem to work
\bibliographystyle{plain} % or use plainnat, unsrt, etc.
\bibliography{termal}

